Series of Lite-On Light (ALS) and Proximity(PS) sensors - LTR-303, LTR-329, LTR-553, LTR-556, LTR-659, some others (#3528)
OttoWinter/esphomedocs: components/sensor/ltr_als_ps.rst

> Lite-On Ambient Light & Proximity Sensors
> ==================================================

Reviewer: ```suggestion
Lite-On Ambient Light & Proximity Sensors
=========================================
```

.. seo::
    :description: Instructions for setting up LTR303, LTR329, LTR553, LTR556, LTR559, LTR659 ambient light sensors/proximity sensors with ESPHome.
    :image: ltr303.jpg
    :keywords: LTR-329, LTR-303, LTR-553, LTR-556, LTR-559, LTR-659

.. figure:: images/ltr303-full.jpg
    :align: center
    :width: 90.0%

    LTR-329 and LTR-303 Ambient Light Sensors.

.. figure:: images/ltr303-ui.png
    :align: center
    :width: 60.0%

    LTR-303 Sensor in Home Assistant UI.

> The ``ltr_als_ps`` sensor platform allows you to use family of several LiteOn ambient light and proximity sensors

Reviewer: ```suggestion
The ``ltr_als_ps`` sensor platform allows you to use a range of LiteOn ambient light and proximity sensors
```

with ESPHome. The :ref:`I²C Bus <i2c>` is required to be set up in your configuration for this sensor to work.

The family of sensors includes:

- Ambient Light Sensors **LTR-303ALS** and **LTR-329ALS**
- Integrated Ambient Light and Proximity Sensors **LTR-553ALS**, **LTR-556ALS**, **LTR-559ALS**
- Proximity Sensor **LTR-659ALS** (with PS gain control)

The sensors are very similar and share the same datasheet. First two have default I²C address ``0x29`` and the rest
have default I²C address ``0x23``. The sensors are available on breakout boards from a number of vendors, for 
example `Adafruit`_ or `Pimoroni`_.

Proximity sensors are the same sort of sensors that you find in phones and tablets to disable the screen when you hold

> the device up to your ear. They might be useful for automated turning on or off of displays and control panels. Good
> example is CoreS3 board from M5Stack that has LTR-559ALS sensor integrated just under the screen.

Reviewer: ```suggestion
the device up to your ear. They might be useful for automated turning on or off of displays and control panels. An
example is the CoreS3 board from M5Stack that has an LTR-559ALS sensor integrated just under the screen.
```

.. _Adafruit: http://www.adafruit.com/products/5610
.. _Pimoroni: https://shop.pimoroni.com/products/ltr-559-light-proximity-sensor-breakout

Ambient light sensing
---------------------

> Sensor platform provides a linear response over a wide dynamic range from 0.01 lux to 64k lux and is well suited 

Reviewer: ```suggestion
The sensor platform provides a linear response over a wide dynamic range from 0.01 lux to 64k lux and is well suited 
```

to applications under high ambient brightness. There are altogether six gain settings (1X, 2X, 4X, 8X, 48X and 96X)
available for user to configure. Use higher gain for dimmer areas.

> Devices consists of two photodiodes: *CH0* diode that is sensitive to both visible and infrared light and 
> *CH1* diode that is sensitive only to infrared light.

Reviewer: ```suggestion
The devices consist of two photodiodes: The *CH0* diode is sensitive to both visible and infrared light and 
the *CH1* diode is sensitive only to infrared light.
```

.. figure:: images/ltr303-spectral.png
    :align: center
    :width: 100.0%

>     CH0 and CH1 channels typical spectral response

Reviewer: ```suggestion
    CH0 and CH1 typical spectral response
```

    IF (RATIO < 0.45)
        ALS_LUX = (1.7743 * CH0 + 1.1059 * CH1) / ALS_GAIN / ALS_INT
    ELSEIF (RATIO < 0.64 && RATIO >= 0.45)
        ALS_LUX = (4.2785 * CH0 – 1.9548 * CH1) / ALS_GAIN / ALS_INT
    ELSEIF (RATIO < 0.85 && RATIO >= 0.64)
        ALS_LUX = (0.5926 * CH0 + 0.1185 * CH1) / ALS_GAIN / ALS_INT
    ELSE
        ALS_LUX = 0
    END
  

where:

- ``CH0`` and ``CH1`` are the sensor values (measurement counts) for Visible + IR (Ch0) and IR only (Ch1) sensors respectively.
- ``ALS_GAIN`` is the gain multiplier
- ``ALS_INT`` is the integration time in ms/100

ALS Gain levels
^^^^^^^^^^^^^^^

The table lists gain values and corresponding illuminance range:

 ========= ================================
  Gain      Illuminance range
 ========= ================================
  ``1X``    1 lux to 64k lux (default)
  ``2X``    0.5 lux to 32k lux
  ``4X``    0.25 lux to 16k lux
  ``8X``    0.125 lux to 8k lux
  ``48X``   0.02 lux to 1.3k lux
  ``96X``   0.01 lux to 600 lux
 ========= ================================

This Wikipedia `article <https://en.wikipedia.org/wiki/Lux>`__ has a table of some lux values for comparison.

Proximity sensing
-----------------

> Proximity sensor has built-in emitter and detector. The sensor detects reflected IR light from the emitter and
> gives a raw count value inversely exponential to the distance. Drop in the count value means an object is getting
> further away from the sensor, and vice versa. Neither of the datasheets provides any information on how to convert
> the raw count value to distance. The only way to do it is to test the sensor yourself and select the threshold
> according to your needs and environment. Exact numbers will depend on the type of the object, its color and 
> reflectivity.

Reviewer: ```suggestion
The proximity sensor has a built-in emitter and detector. The sensor detects reflected IR light from the emitter and
gives a raw count value inversely exponential to the distance. A decrease in the count value means an object is getting
further away from the sensor, and vice-versa. Neither of the datasheets provides any information on how to convert
the raw count value to distance. The only way to do it is to test the sensor yourself and select the threshold
according to your needs and environment. Exact numbers will depend on the type of the object, its color and 
reflectivity.
```

Example configuration
---------------------

.. code-block:: yaml

>     i2c:
>       # ...
> 

Reviewer: ```suggestion
```

- **update_interval** (*Optional*, :ref:`config-time`): The interval for checking the sensors.
  Defaults to ``60s``.
- **ps_cooldown** (*Optional*, :ref:`config-time`): The "cooldown" period after the proximity sensor is triggered. 
  Helps to avoid multiple calls.  Defaults to ``5s``.
- **ps_gain** (*Optional*, string): The gain the device will use for proximity sensor. Higher values are better in low-light conditions.
  Valid values are ``16X`` *(default)*, ``32X``, ``64X``. Only for **LTR-659ALS**.
- **ps_high_threshold** (*Optional*, int): The threshold for the proximity sensor to trigger on object getting closer. 
  Defaults to ``65535``, which implies it will never be triggered.
- **ps_low_threshold** (*Optional*, int): The threshold for the proximity sensor to trigger on object getting further away. 
  Defaults to ``0``, which implies it will never be triggered.
- **on_ps_high_threshold** (*Optional*): Actions to perform when the proximity sensor is triggered
  on object getting closer.
- **on_ps_low_threshold** (*Optional*): Actions to perform when the proximity sensor is triggered
  on object getting further away.

Sensors
^^^^^^^

> This component offers five sensors for ALS-enabled devices and one sensor for PS-enabled devices.
> You can configure all or any subset of the sensors. Each configured sensor is reported separately 
> on each ``update_interval``. **name** option is required for the sensor. All other options from 
> :ref:`Sensor <config-sensor>`. If you don’t need to configure any other sensor options, you can use 
> the simplified syntax for the sensor: ``ambient_light: "Ambient light"``, for example.

Reviewer: ```suggestion
This component offers five sensors for ALS-equipped devices and one sensor for PS-equipped devices.
You can configure all or any subset of these sensors. Each configured sensor is reported separately 
on each ``update_interval``. Each is an ESPHome :ref:`sensor <config-sensor>` and may be configured
accordingly; if you don’t need to configure additional :ref:`sensor <config-sensor>` variables, you may simply
use the shorthand syntax for the sensor. For example: ``ambient_light: "Ambient light"``
```

> accordingly; if you don’t need to configure additional :ref:`sensor <config-sensor>` variables, you may simply
> use the shorthand syntax for the sensor. For example: ``ambient_light: "Ambient light"``

Reviewer: ```suggestion
accordingly; if you don’t need to configure additional :ref:`sensor <config-sensor>` variables, you
may simply use the shorthand syntax for the sensor. For example: ``ambient_light: "Ambient light"``
```
